File: app/src/main/java/com/codelabs/accountsafetyfido/java/HomeAct.java
/*
 *  Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.codelabs.accountsafetyfido.java;

import android.os.Bundle;
import android.text.TextUtils;
import android.util.Base64;
import android.util.Log;
import android.view.View;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.appcompat.app.AppCompatActivity;
import androidx.appcompat.widget.AppCompatImageView;
import androidx.appcompat.widget.AppCompatTextView;
import androidx.appcompat.widget.LinearLayoutCompat;
import androidx.appcompat.widget.SwitchCompat;
import androidx.core.content.ContextCompat;

import com.bumptech.glide.Glide;
import com.codelabs.accountsafetyfido.R;
import com.codelabs.accountsafetyfido.java.bean.UserBean;
import com.codelabs.accountsafetyfido.java.common.SPConstants;
import com.codelabs.accountsafetyfido.java.util.SPUtil;
import com.codelabs.accountsafetyfido.java.util.UserUtil;
import com.codelabs.accountsafetyfido.java.util.fingerprint.FingerprintHelper;
import com.codelabs.accountsafetyfido.java.util.fingerprint.KeyHelper;
import com.codelabs.accountsafetyfido.java.weight.FpCloseDialogFg;
import com.codelabs.accountsafetyfido.java.weight.FpOpenDialogFg;
import com.codelabs.accountsafetyfido.java.weight.IFingerprintListener;
import com.huawei.hmf.tasks.OnFailureListener;
import com.huawei.hmf.tasks.OnSuccessListener;
import com.huawei.hmf.tasks.Task;
import com.huawei.hms.support.api.fido.bioauthn.BioAuthnCallback;
import com.huawei.hms.support.api.fido.bioauthn.BioAuthnManager;
import com.huawei.hms.support.api.fido.bioauthn.BioAuthnPrompt;
import com.huawei.hms.support.api.fido.bioauthn.BioAuthnResult;
import com.huawei.hms.support.api.fido.bioauthn.CryptoObject;
import com.huawei.hms.support.hwid.HuaweiIdAuthManager;
import com.huawei.hms.support.hwid.request.HuaweiIdAuthParams;
import com.huawei.hms.support.hwid.request.HuaweiIdAuthParamsHelper;
import com.huawei.hms.support.hwid.service.HuaweiIdAuthService;

import javax.crypto.Cipher;

/**
 * Created by CodeLab
 */
public class HomeAct extends AppCompatActivity {
    private static final String TAG = "--HomeAct--";

    // avatar imageView
    private AppCompatImageView mIvAvatar;
    // display textView
    private AppCompatTextView mTvDisplayName;
    // openId textView
    private AppCompatTextView mTvOpenId;
    // Fingerprint content
    private LinearLayoutCompat mLlFingerprint;
    // Fingerprint switch button
    private SwitchCompat mSwitchCompat;

    @Override
    protected void onCreate(@Nullable Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.home_act);
        onInitView();
        onInitialize();
    }

    /**
     * init view
     */
    private void onInitView() {
        mIvAvatar = findViewById(R.id.home_avatar);
        mTvDisplayName = findViewById(R.id.home_displayName);
        mTvOpenId = findViewById(R.id.hone_openid);
        mLlFingerprint = findViewById(R.id.home_fingerprint_content);
        mSwitchCompat = findViewById(R.id.home_fingerprint);
        findViewById(R.id.home_exit).setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                onExitLogin();
            }
        });
    }

    /**
     * init data
     */
    private void onInitialize() {
        String openId = (String) SPUtil.get(this, SPConstants.KEY_OPENID, "");
        if (TextUtils.isEmpty(openId)) {
            finish();
            return;
        }
        mTvOpenId.setText(openId);
        UserBean userBean = UserUtil.getLocalUser(this, openId);
        if (userBean == null) {
            return;
        }
        mTvDisplayName.setText(userBean.getDisplayName());
        // avatar
        Glide.with(this)
                .asBitmap()
                .skipMemoryCache(true)
                .load(userBean.getAvatar())
                .error(R.mipmap.ic_launcher_round)
                .into(mIvAvatar);

        onFingerprintCheck();
    }

    /**
     * Fingerprint Check
     */
    private void onFingerprintCheck() {
        // check fingerprint is support
        BioAuthnManager bioAuthnManager = new BioAuthnManager(this);
        int errorCode = bioAuthnManager.canAuth();
        if (errorCode != 0) {
            // not support
            mLlFingerprint.setVisibility(View.GONE);
            SPUtil.put(HomeAct.this, SPConstants.FINGER_PRINT_LOGIN_SWITCH, false);
            return;
        }
        mLlFingerprint.setVisibility(View.VISIBLE);
        boolean isOpen = (boolean) SPUtil.get(HomeAct.this, SPConstants.FINGER_PRINT_LOGIN_SWITCH, false);
        mSwitchCompat.setChecked(isOpen);
        mSwitchCompat.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                if (mSwitchCompat.isChecked()) {
                    // open fingerprint
                    onFingerprintOpenCheck();
                } else {
                    // close fingerprint
                    onFingerprintCloseCheck();
                }
            }
        });
    }

    /**
     * open fingerprint check
     */
    private void onFingerprintOpenCheck() {
        FpOpenDialogFg fpOpenDialogFg = FpOpenDialogFg.newInstance();
        fpOpenDialogFg.setListener(new IFingerprintListener() {
            @Override
            public void onFingerprintListener(int type) {
                if (type == 0) {
                    onFingerprintOpenResult(false);
                } else {
                    onFingerprintOpen();
                }
            }
        });
        fpOpenDialogFg.show(getSupportFragmentManager(), "fpOpenDialogFg");
    }

    /**
     * open fingerprint
     */
    private void onFingerprintOpen() {
        BioAuthnPrompt bioAuthnPrompt = new BioAuthnPrompt(this, ContextCompat.getMainExecutor(this), authCallback);
        BioAuthnPrompt.PromptInfo.Builder builder = new BioAuthnPrompt.PromptInfo.Builder()
                .setTitle("This is the title.")
                .setSubtitle("This is the subtitle")
                .setDescription("This is the description");
        builder.setNegativeButtonText("This is the 'Cancel' button.");
        BioAuthnPrompt.PromptInfo info = builder.build();
        // get cryptoObject
        CryptoObject cryptoObject = new CryptoObject(KeyHelper.getInstance().getEncryptCipher());
        bioAuthnPrompt.auth(info, cryptoObject);
    }

    /**
     * open fingerprint result
     */
    private void onFingerprintOpenResult(boolean isSuccess) {
        SPUtil.put(HomeAct.this, SPConstants.FINGER_PRINT_LOGIN_SWITCH, isSuccess);
        mSwitchCompat.setChecked(isSuccess);
    }

    /**
     * close fingerprint check
     */
    private void onFingerprintCloseCheck() {
        FpCloseDialogFg fpCloseDialogFg = FpCloseDialogFg.newInstance();
        fpCloseDialogFg.setListener(new IFingerprintListener() {
            @Override
            public void onFingerprintListener(int type) {
                if (type == 0) {
                    mSwitchCompat.setChecked(true);
                } else {
                    onCloseFingerprint();
                }
            }
        });
        fpCloseDialogFg.show(getSupportFragmentManager(), "fpCloseDialogFg");
    }

    /**
     * close fingerprint
     */
    private void onCloseFingerprint() {
        FingerprintHelper.delete(this, SPConstants.KEY_SAVE_IV);
        FingerprintHelper.delete(this, SPConstants.KEY_SAVE_ENCODE);
        SPUtil.remove(HomeAct.this, SPConstants.FINGER_PRINT_LOGIN_SWITCH);
        mSwitchCompat.setChecked(false);
    }

    /**
     * Exit Login
     */
    private void onExitLogin() {
        HuaweiIdAuthParams authParams = new HuaweiIdAuthParamsHelper(HuaweiIdAuthParams.DEFAULT_AUTH_REQUEST_PARAM)
                .setIdToken().createParams();
        HuaweiIdAuthService mAuthManager = HuaweiIdAuthManager.getService(this, authParams);

        Task<Void> signOutTask = mAuthManager.signOut();
        signOutTask.addOnSuccessListener(new OnSuccessListener<Void>() {
            @Override
            public void onSuccess(Void aVoid) {
                Log.i(TAG, "signOut Success");
                onClearExitData();
            }
        }).addOnFailureListener(new OnFailureListener() {
            @Override
            public void onFailure(Exception e) {
                Log.e(TAG, "signOut fail");
                onClearExitData();
            }
        });
    }

    /**
     * clear exit data
     */
    private void onClearExitData() {
        String openId = (String) SPUtil.get(this, SPConstants.KEY_OPENID, "");
        SPUtil.remove(this, openId);
        SPUtil.remove(this, SPConstants.KEY_OPENID);
        finish();
    }

    // call back
    BioAuthnCallback authCallback = new BioAuthnCallback() {
        @Override
        public void onAuthError(int errMsgId, @NonNull CharSequence errString) {
            // TODO Auth Error
            Log.e(TAG, "auth error : " + errString);
            onFingerprintOpenResult(false);
        }

        @Override
        public void onAuthSucceeded(@NonNull BioAuthnResult result) {
            // Auth Success
            try {
                CryptoObject cryptoObject = result.getCryptoObject();
                if (null == cryptoObject) {
                    Log.e(TAG, "auth success cryptoObject is null");
                    onFingerprintOpenResult(false);
                    return;
                }
                Cipher cipher = result.getCryptoObject().getCipher();
                if (null == cipher) {
                    Log.e(TAG, "auth success cipher is null");
                    onFingerprintOpenResult(false);
                    return;
                }
                String openId = (String) SPUtil.get(HomeAct.this, SPConstants.KEY_OPENID, "");
                byte[] bytes = cipher.doFinal(openId.getBytes());
                // save Encode
                FingerprintHelper.put(HomeAct.this, SPConstants.KEY_SAVE_ENCODE, Base64.encodeToString(bytes, Base64.URL_SAFE));
                // save IV
                byte[] iv = cipher.getIV();
                FingerprintHelper.put(HomeAct.this, SPConstants.KEY_SAVE_IV, Base64.encodeToString(iv, Base64.URL_SAFE));
                Log.i(TAG, "auth success");
                onFingerprintOpenResult(true);
            } catch (Exception e) {
                e.printStackTrace();
            }
        }

        @Override
        public void onAuthFailed() {
            // TODO Auth Failed
            Log.e(TAG, "auth failed");
            onFingerprintOpenResult(false);
        }
    };
}

File: app/src/main/java/com/codelabs/accountsafetyfido/kotlin/HomeAct.kt
/*
 *  Copyright 2020. Huawei Technologies Co., Ltd. All rights reserved.
 *
 *  Licensed under the Apache License, Version 2.0 (the "License");
 *  you may not use this file except in compliance with the License.
 *  You may obtain a copy of the License at
 *
 *    http://www.apache.org/licenses/LICENSE-2.0
 *
 *  Unless required by applicable law or agreed to in writing, software
 *  distributed under the License is distributed on an "AS IS" BASIS,
 *  WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 *  See the License for the specific language governing permissions and
 *  limitations under the License.
 */
package com.codelabs.accountsafetyfido.kotlin

import android.os.Bundle
import android.text.TextUtils
import android.util.Base64
import android.util.Log
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.core.content.ContextCompat
import com.bumptech.glide.Glide
import com.codelabs.accountsafetyfido.R
import com.codelabs.accountsafetyfido.kotlin.bean.UserBean
import com.codelabs.accountsafetyfido.kotlin.common.SPConstants
import com.codelabs.accountsafetyfido.kotlin.util.SPUtil
import com.codelabs.accountsafetyfido.kotlin.util.UserUtil
import com.codelabs.accountsafetyfido.kotlin.util.fingerprint.FingerprintHelper
import com.codelabs.accountsafetyfido.kotlin.util.fingerprint.KeyHelper
import com.codelabs.accountsafetyfido.kotlin.weight.FpCloseDialogFg
import com.codelabs.accountsafetyfido.kotlin.weight.FpOpenDialogFg
import com.codelabs.accountsafetyfido.kotlin.weight.IFingerprintListener
import com.huawei.hms.support.api.fido.bioauthn.*
import com.huawei.hms.support.hwid.HuaweiIdAuthManager
import com.huawei.hms.support.hwid.request.HuaweiIdAuthParams
import com.huawei.hms.support.hwid.request.HuaweiIdAuthParamsHelper
import kotlinx.android.synthetic.main.home_act.*

/**
 * Created by CodeLab
 */
class HomeAct : AppCompatActivity() {

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.home_act)
        home_exit.setOnClickListener { onExitLogin() }
        onInitialize()
    }

    /**
     * init data
     */
    private fun onInitialize() {
        val openId = SPUtil[this, SPConstants.KEY_OPENID, ""] as String
        if (TextUtils.isEmpty(openId)) {
            finish()
            return
        }
        hone_openid.text = openId
        val userBean: UserBean? = UserUtil.getLocalUser(this, openId)
        home_displayName.text = userBean!!.displayName
        // avatar
        Glide.with(this).asBitmap().skipMemoryCache(true)
            .load(userBean.avatar).error(R.mipmap.ic_launcher_round)
            .into(home_avatar)
        onFingerprintCheck()
    }

    /**
     * Fingerprint Check
     */
    private fun onFingerprintCheck() {
        // check fingerprint is support
        val bioAuthnManager = BioAuthnManager(this)
        val errorCode = bioAuthnManager.canAuth()
        if (errorCode != 0) {
            // not support
            home_fingerprint_content.visibility = View.GONE
            SPUtil.put(this, SPConstants.FINGER_PRINT_LOGIN_SWITCH, false)
            return
        }
        val isOpen = SPUtil[this, SPConstants.FINGER_PRINT_LOGIN_SWITCH, false] as Boolean
        home_fingerprint.isChecked = isOpen

        home_fingerprint.setOnClickListener {
            if (home_fingerprint.isChecked) {
                // open fingerprint
                onFingerprintOpenCheck();
            } else {
                // close fingerprint
                onFingerprintCloseCheck();
            }
        }
    }

    /**
     * open fingerprint check
     */
    private fun onFingerprintOpenCheck() {
        val fpOpenDialogFg = FpOpenDialogFg.newInstance()
        fpOpenDialogFg.setListener(object : IFingerprintListener {
            override fun onFingerprintListener(type: Int) {
                if (type == 0) {
                    onFingerprintOpenResult(false);
                } else {
                    onFingerprintOpen()
                }
            }
        })
        fpOpenDialogFg.show(supportFragmentManager, "fpOpenDialogFg")
    }

    /**
     * open fingerprint result
     */
    private fun onFingerprintOpenResult(isSuccess: Boolean) {
        SPUtil.put(this, SPConstants.FINGER_PRINT_LOGIN_SWITCH, isSuccess)
        home_fingerprint.isChecked = isSuccess
    }

    /**
     * close fingerprint check
     */
    private fun onFingerprintCloseCheck() {
        val fpCloseDialogFg = FpCloseDialogFg.newInstance()
        fpCloseDialogFg.setListener(object : IFingerprintListener {
            override fun onFingerprintListener(type: Int) {
                if (type == 0) {
                    home_fingerprint.isChecked = true
                } else {
                    onCloseFingerprint()
                }
            }
        })
        fpCloseDialogFg.show(supportFragmentManager, "fpCloseDialogFg")
    }

    /**
     * close fingerprint
     */
    private fun onCloseFingerprint() {
        FingerprintHelper.delete(this, SPConstants.KEY_SAVE_IV)
        FingerprintHelper.delete(this, SPConstants.KEY_SAVE_ENCODE)
        SPUtil.remove(this, SPConstants.FINGER_PRINT_LOGIN_SWITCH)
        home_fingerprint.isChecked = false
    }

    /**
     * open fingerprint
     */
    private fun onFingerprintOpen() {
        val bioAuthnPrompt = BioAuthnPrompt(this, ContextCompat.getMainExecutor(this), authCallback)
        val builder = BioAuthnPrompt.PromptInfo.Builder()
            .setTitle("This is the title.")
            .setSubtitle("This is the subtitle")
            .setDescription("This is the description")
        builder.setNegativeButtonText("This is the 'Cancel' button.")
        val info = builder.build()
        // get cryptoObject
        val cryptoObject = KeyHelper.getInstance().getEncryptCipher()?.let { CryptoObject(it) }
        if (cryptoObject != null) {
            bioAuthnPrompt.auth(info, cryptoObject)
        }
    }

    /**
     * Exit Login
     */
    private fun onExitLogin() {
        val authParams = HuaweiIdAuthParamsHelper(HuaweiIdAuthParams.DEFAULT_AUTH_REQUEST_PARAM)
            .setIdToken().createParams()
        val mAuthManager = HuaweiIdAuthManager.getService(this, authParams)
        val signOutTask = mAuthManager!!.signOut()
        signOutTask.addOnSuccessListener {
            Log.i("exitLogin", "signOut Success")
            onClearExitData()
        }.addOnFailureListener {
            Log.e("exitLogin", "signOut fail")
            onClearExitData()
        }
    }

    /**
     * clear exit data
     */
    private fun onClearExitData() {
        val openId = SPUtil[this, SPConstants.KEY_OPENID, ""] as String
        SPUtil.remove(this, openId)
        SPUtil.remove(this, SPConstants.KEY_OPENID)
        finish()
    }

    // call back
    private var authCallback: BioAuthnCallback = object : BioAuthnCallback() {
        override fun onAuthError(errMsgId: Int, errString: CharSequence) {
            // TODO Auth Error
            Log.e("authCallback", "auth error : $errString")
            onFingerprintOpenResult(false);
        }

        override fun onAuthSucceeded(result: BioAuthnResult) {
            // Auth Success
            try {
                val cryptoObject = result.cryptoObject
                if (null == cryptoObject) {
                    Log.e("authCallback", "auth success cryptoObject is null")
                    onFingerprintOpenResult(false);
                    return
                }
                val cipher = result.cryptoObject!!.cipher
                if (null == cipher) {
                    Log.e("authCallback", "auth success cipher is null")
                    onFingerprintOpenResult(false);
                    return
                }
                val openId = SPUtil[this@HomeAct, SPConstants.KEY_OPENID, ""] as String
                val bytes = cipher.doFinal(openId.toByteArray())
                // save Encode
                FingerprintHelper.put(this@HomeAct, SPConstants.KEY_SAVE_ENCODE, Base64.encodeToString(bytes, Base64.URL_SAFE))
                // save IV
                val iv = cipher.iv
                FingerprintHelper.put(this@HomeAct, SPConstants.KEY_SAVE_IV, Base64.encodeToString(iv, Base64.URL_SAFE))
                Log.i("authCallback", "auth success")
                onFingerprintOpenResult(true);
            } catch (e: Exception) {
                e.printStackTrace()
            }
        }

        override fun onAuthFailed() {
            // TODO Auth Failed
            Log.e("authCallback", "auth failed")
            onFingerprintOpenResult(false);
        }
    }
}